buster/validators/question_answer_validator.py

```diff
@@ -1,90 +1,133 @@
-
+class QuestionValidator:
+    def __init__(self, completion_kwargs: dict, check_question_prompt: str, invalid_question_response: str):
         self.completer = ChatGPTCompleter(completion_kwargs=completion_kwargs)
         self.check_question_prompt = check_question_prompt
-        self.unknown_response_templates = unknown_response_templates
+        self.invalid_question_response = invalid_question_response
 
     def check_question_relevance(self, question: str) -> tuple[bool, str]:
-        """Determines wether a question is relevant or not for our given framework."""
-
-        def get_relevance(outputs: str) -> bool:
-            # remove trailing periods, happens sometimes...
-            outputs = outputs.strip(".").lower()
-
-            if outputs == "true":
-                relevance = True
-            elif outputs == "false":
-                relevance = False
-            else:
-                # Default assume it's no longer relevant if the detector didn't give one of [true, false]
-                logger.warning(f"the question validation returned an unexpeced value: {outputs}. Assuming Invalid...")
-                relevance = False
-            return relevance
-
-        response = self.invalid_question_response
+        """Determines whether a question is relevant for our given framework."""
         try:
-            outputs, error = self.completer.complete(self.check_question_prompt, user_input=question)
-            relevance = get_relevance(outputs)
+            outputs, _ = self.completer.complete(self.check_question_prompt, user_input=question)
+            outputs = outputs.strip(".").lower()
+            if outputs not in ["true", "false"]:
+                logger.warning(f"the question validation returned an unexpeced value: {outputs=}. Assuming Invalid...")
+            relevance = outputs.strip(".").lower() == "true"
+            response = self.invalid_question_response
 
         except Exception as e:
-            # Something went wrong, assume immediately not relevant.
-            logger.exception("Something went wrong during question relevance detection. See traceback:")
+            logger.exception("Error during question relevance detection.")
             relevance = False
             response = "Unable to process your question at the moment, try again soon"
 
-        logger.info(f"Question {relevance=}")
-
         return relevance, response
 
-    def check_answer_relevance(self, answer: str) -> bool:
-        """Check to see if a generated answer is relevant to the chatbot's knowledge or not.
 
-        We assume we've prompt-engineered our bot to say a response is unrelated to the context if it isn't relevant.
-        Here, we compare the embedding of the response to the embedding of the prompt-engineered "I don't know" embedding.
+class AnswerValidator:
+    def __init__(self, unknown_response_templates: list[str], unknown_threshold: float, embedding_fn: callable = None):
```

Comment:
I hadn't realised before but `callable` is not a typehint, it's the actual python keyword. The proper typehint is `Callable`, and in this case you could do something like `Callable[[str], np.ndarray]`